### PR TITLE
support config istiod server tls cipher suites

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
### Context

Sometimes it is not allowed to support tls1.0/tls1.1 and also not allowed to use some cipher suites in production env.  So in this pr, I make cipher suites of istiod TLS server configurable.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
https://github.com/istio/api/pull/1912 - There is already an API under discussion

File: pilot/pkg/bootstrap/webhook.go [hunk 1]
```diff
@@ -41,11 +41,14 @@ func (s *Server) initSecureWebhookServer(args *PilotArgs) {
 	log.Info("initializing secure webhook server for istiod webhooks")
 	// create the https server for hosting the k8s injectionWebhook handlers.
 	s.httpsMux = http.NewServeMux()
+	cipherSuites, _ := TLSCipherSuites(args.ServerOptions.TLSOptions.TLSCipherSuites)
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/webhook.go:44]: we probably shouldn't ignore the error?

File: pilot/cmd/pilot-discovery/options.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	for _, cipher := range tls.InsecureCipherSuites() {
+		cipherKeys.Insert(cipher.Name)
+	}
+	return cipherKeys.List()
+}
+
+// secureTLSCipherNames returns a list of cipher suite names implemented by crypto/tls.
+func secureTLSCipherNames() []string {
+	cipherKeys := sets.NewString()
+	for _, cipher := range tls.CipherSuites() {
```
Comment (howardjohn) [inline pilot/cmd/pilot-discovery/options.go:38]: `CipherSuites()` contains *all* of them, not "secure" ones.

**Review by howardjohn — COMMENTED** [review]
The general idea looks good. @lei-tang has been looking at the same thing but for sidecar-sidecar communication so it would be good for him to take a look as well

File: pilot/cmd/pilot-discovery/options.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+// which have security issues.
+func insecureTLSCipherNames() []string {
+	cipherKeys := sets.NewString()
+	for _, cipher := range tls.InsecureCipherSuites() {
+		cipherKeys.Insert(cipher.Name)
+	}
+	return cipherKeys.List()
+}
+
+// secureTLSCipherNames returns a list of cipher suite names implemented by crypto/tls.
```
Comment (lei-tang) [inline pilot/cmd/pilot-discovery/options.go:35]: Based on the function name, the comment "secureTLSCipherNames returns a list of cipher suite names implemented by crypto/tls" should be "secureTLSCipherNames returns a list of secure cipher suites implemented by crypto/tls" instead.

File: pilot/cmd/pilot-discovery/main.go [hunk 4]
```diff
@@ -154,6 +158,11 @@ func init() {
 		"File containing the x509 Server Certificate")
 	discoveryCmd.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.KeyFile, "tlsKeyFile", "",
 		"File containing the x509 private key matching --tlsCertFile")
+	discoveryCmd.PersistentFlags().StringSliceVar(&serverArgs.ServerOptions.TLSOptions.TLSCipherSuites, "tls-cipher-suites", nil,
+		"Comma-separated list of cipher suites for the server. "+
```
Comment (lei-tang) [inline pilot/cmd/pilot-discovery/main.go:162]: For better code readability, please change "Comma-separated list of cipher suites for the server" to "Comma-separated list of cipher suites for the istiod TLS server".
In addition, please add an example configuration for tls-cipher-suites.

**Comment by lei-tang** [comment]
For better readability of the PR, can you update the PR description to mention that the PR is to make the cipher suites of istiod TLS server configurable? For example, change 
"So in this pr, i make cipher suites configurable" to "So in this pr, I make cipher suites of istiod TLS server configurable".

**Review by lei-tang — CHANGES_REQUESTED** [review]
Please add tests (unit tests and integration tests). To verify that the cipher suites configured for the istiod TLS server work properly, the integration tests may:
- Create a client to connect to the istiod TLS server using the cipher suites configured and verify that the TLS connection succeeds.
- Create a client to connect to the istiod TLS server using a cipher not configured for the istiod TLS server and verify that the TLS connection fails.

File: pilot/cmd/pilot-discovery/options.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-discovery/options.go:38]: Nope, if look closely at it.  It is totally different from tls.InsecureCipherSuites

File: pilot/pkg/bootstrap/webhook.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/webhook.go:44]: OK, i should look for another way to parse it outside and paas in the result.

File: pilot/cmd/pilot-discovery/options.go (hunk 2 above)
Comment (howardjohn) [reply pilot/cmd/pilot-discovery/options.go:38]: Ah, I read it wrong. I think this list is still larger than we would like? We picked a much smaller set of sidecar communication

File: pilot/cmd/pilot-discovery/main.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-discovery/main.go:162]: If user run `cmd -h`. It will tell him how to configure.

**Comment by hzxuzhonghu** [comment]
@lei-tang IMO, this can be covered by ut. Actually, this function has been tested by golang tls lib.

**Comment by lei-tang** [comment]
> @lei-tang IMO, this can be covered by ut. Actually, this function has been tested by golang tls lib.

The PR adds a feature to configure TLS cipher suites for istiod server. The tests in standard golang TLS library is orthogonal to the feature in this PR. In other words, the passing of tests in standard golang TLS library does not mean that the feature in this PR work as expected. 

Without integration tests, it is uncertain whether the feature in this PR works or not. There may be bugs such as 1) a user may configure cipher-1 and cipher-2 for the istiod TLS server but a TLS client using cipher-1 is unable to connect to the istiod TLS server. 2) a user may configure cipher-1 and cipher-2 for the istiod TLS server but a TLS client using cipher-3 also successfully connects to the istiod TLS server, which should never be allowed.

**Comment by howardjohn** [comment]
personally, an integration test seems overkill here and unnecessarily high bar for this PR

**Comment by hzxuzhonghu** [comment]
This is a normal tls server-client config issue. Not tightly related to istio. So I think it is not necessary to enforce an integration test for it.

**Comment by hzxuzhonghu** [comment]
I can add a ut and simulate c-s communication over tls.

**Comment by lei-tang** [comment]
If no integration tests for this feature, how to automate the testing of the feature to verify that the feature does not have bugs such as the following two examples?

1) a user may configure cipher-1 and cipher-2 for the istiod TLS server but a TLS client using cipher-1 is unable to connect to the istiod TLS server. 
2) a user may configure cipher-1 and cipher-2 for the istiod TLS server but a TLS client using cipher-3 also successfully connects to the istiod TLS server, which should never be allowed.

**Comment by hzxuzhonghu** [comment]
Your assumption is totally based on tls lib does not work well.

File: pilot/pkg/bootstrap/server_test.go [hunk 5]
```diff
@@ -330,6 +332,124 @@ func TestNewServer(t *testing.T) {
...
+
+			// wait for the https server start
+			time.Sleep(time.Second)
+
+			httpsReadyClient := &http.Client{
+				Timeout: time.Second,
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: true,
+						CipherSuites:       c.clientCipherSuites,
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server_test.go:422]: @lei-tang I think this is what you want.

**Comment by lei-tang** [comment]
> Your assumption is totally based on tls lib does not work well.

The standard golang TLS library should work well. The feature in this PR should likely work well too. However, whether a feature works or not should not be based on assumptions but based on the actual tests.

**Comment by hzxuzhonghu** [comment]
Another question, should we add a test for 1+1=2? According to your theory, this is a must.

File: pilot/pkg/bootstrap/server_test.go (hunk 5 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/server_test.go:422]: Thanks for adding tests! The test looks good.

**Comment by lei-tang** [comment]
Thanks for adding tests! The test looks good and I will remove my hold.

**Review by ramaraochavali — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
Thanks @all for your comments.

File: pilot/pkg/bootstrap/webhook.go [hunk 6]
```diff
@@ -31,8 +32,9 @@ const (
 // If https address is off the injection handlers will be registered on the main http endpoint, with
 // TLS handled by a proxy/gateway in front of Istiod.
 func (s *Server) initSecureWebhookServer(args *PilotArgs) {
+	fmt.Println(args.ServerOptions.HTTPSAddr)
 	// create the https server for hosting the k8s injectionWebhook handlers.
-	if s.kubeClient == nil || args.ServerOptions.HTTPSAddr == "" {
+	if args.ServerOptions.HTTPSAddr == "" {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/webhook.go:37]: Why is this changed?

File: pilot/pkg/bootstrap/webhook.go [hunk 7]
```diff
@@ -31,8 +32,9 @@ const (
 // If https address is off the injection handlers will be registered on the main http endpoint, with
 // TLS handled by a proxy/gateway in front of Istiod.
 func (s *Server) initSecureWebhookServer(args *PilotArgs) {
+	fmt.Println(args.ServerOptions.HTTPSAddr)
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/webhook.go:35]: Please remove?

**Review by howardjohn — COMMENTED** [review]
@hzxuzhonghu can you revert part of this? It looks unrelated
